cmd/otel-allocator/allocation/allocator.go

```diff
@@ -0,0 +1,310 @@
+	defer timer.ObserveDuration()
+
+	CollectorsAllocatable.WithLabelValues(a.strategy.GetName()).Set(float64(len(collectors)))
+	if len(collectors) == 0 {
+		a.log.Info("No collector instances present")
+	}
+
+	a.m.Lock()
+	defer a.m.Unlock()
+
+	// Check for collector changes
+	collectorsDiff := diff.Maps(a.collectors, collectors)
+	if len(collectorsDiff.Additions()) != 0 || len(collectorsDiff.Removals()) != 0 {
+		a.handleCollectors(collectorsDiff)
+	}
+}
+
+func (a *allocator) GetTargetsForCollectorAndJob(collector string, job string) []*target.Item {
+	a.m.RLock()
+	defer a.m.RUnlock()
+	if _, ok := a.targetItemsPerJobPerCollector[collector]; !ok {
+		return []*target.Item{}
+	}
+	if _, ok := a.targetItemsPerJobPerCollector[collector][job]; !ok {
+		return []*target.Item{}
+	}
+	targetItemsCopy := make([]*target.Item, len(a.targetItemsPerJobPerCollector[collector][job]))
+	index := 0
+	for targetHash := range a.targetItemsPerJobPerCollector[collector][job] {
+		targetItemsCopy[index] = a.targetItems[targetHash]
+		index++
+	}
+	return targetItemsCopy
+}
+
+// TargetItems returns a shallow copy of the targetItems map.
+func (a *allocator) TargetItems() map[string]*target.Item {
+	a.m.RLock()
+	defer a.m.RUnlock()
+	targetItemsCopy := make(map[string]*target.Item)
+	for k, v := range a.targetItems {
+		targetItemsCopy[k] = v
+	}
+	return targetItemsCopy
+}
+
+// Collectors returns a shallow copy of the collectors map.
+func (a *allocator) Collectors() map[string]*Collector {
+	a.m.RLock()
+	defer a.m.RUnlock()
+	collectorsCopy := make(map[string]*Collector)
+	for k, v := range a.collectors {
+		collectorsCopy[k] = v
+	}
+	return collectorsCopy
+}
+
+// handleTargets receives the new and removed targets and reconciles the current state.
+// Any removals are removed from the allocator's targetItems and unassigned from the corresponding collector.
+// Any net-new additions are assigned to the collector on the same node as the target.
+func (a *allocator) handleTargets(diff diff.Changes[*target.Item]) {
+	// Check for removals
+	for k, item := range a.targetItems {
+		// if the current item is in the removals list
+		if _, ok := diff.Removals()[k]; ok {
+			a.removeTargetItem(item)
+		}
+	}
+
+	// Check for additions
+	assignmentErrors := []error{}
+	for k, item := range diff.Additions() {
+		// Do nothing if the item is already there
+		if _, ok := a.targetItems[k]; ok {
+			continue
+		} else {
+			// TODO: track target -> collector relationship in a separate map
+			item.CollectorName = ""
+			// Add item to item pool and assign a collector
+			err := a.addTargetToTargetItems(item)
+			if err != nil {
+				assignmentErrors = append(assignmentErrors, err)
+			}
+		}
+	}
+
+	// Check for unassigned targets
+	unassignedTargets := len(assignmentErrors)
+	if unassignedTargets > 0 {
+		err := errors.Join(assignmentErrors...)
+		a.log.Info("Could not assign targets for some jobs due to missing node labels", "targets", unassignedTargets, "error", err)
```

Comment:
it's not necessarily true that this is due to the missing node labels anymore (though in practice that will probably be the case for now)